core/src/main/java/io/questdb/log/LogRollingFileWriter.java

```diff
@@ -244,6 +278,40 @@ private void flush() {
         _wptr = buf;
     }
 
+    private void removeOldLogs(){
+        if(this.lifeDuration != null) {
+            ff.iterateDir(path.of(logDir).$(), removeExpiredLogsVisitor);
+        }
+        if(this.sizeLimit != null) {
+            currentLogSizeSum = 0;
+            ff.iterateDir(path.of(logDir).$(), removeExcessiveLogsVisitor);
+        }
+    }
+    private void removeOldLogsVisitor(long filePointer, int type){
+        path.trimTo(logDir.length());
+        path.concat(filePointer).$();
+        NativeLPSZ name = new NativeLPSZ();
+        name.of(filePointer);
+        if (Files.notDots(filePointer) && type == Files.DT_FILE && name.toString().contains(logFileTemplate)
+            && clock.getTicks() - ff.getLastModified(path.$())*Timestamps.MILLI_MICROS > nLifeDuration) {
+            if(!ff.remove(path)) {
+                throw new LogError("cannot remove: " + path.$());
+            }
+        }
+    }
+
+    private void removeExcessiveLogsVisitor(long filePointer, int type){
```

Comment:
Is this visitor deterministic? E.g. it should always start deleting the oldest files when it runs which doesn't seem to be the case. Also, it looks like it tries to delete all files once the total log directory size is reached even when it's enough to delete a single file to move the total size back under the threshold.

Reply:
In my experience testing, it always seems to loop through the files from newest to oldest. Because of this, I designed it so that it keeps a running total of the file sizes as it loops from newest to oldest. As soon as a file pushes the total past the limit, the file is deleted along with all subsequent files. This should not delete all of the files unless the newest one is larger than the size limit. Also, I'm not 100% certain that it will always loop newest to oldest, but it always has for me.

Reply:
> Also, I'm not 100% certain that it will always loop newest to oldest, but it always has for me.

Under the hood `ff.iterateDir()` is nothing more than a [`readdir(3)`](https://man7.org/linux/man-pages/man3/readdir.3.html) sys call. Here is a fragment of the call documentation:

"The order in which filenames are read by successive calls to readdir() depends on the filesystem implementation; it is   unlikely that the names will be sorted in any fashion."

On Windows, a [`FindNextFileW`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findnextfilew) sys call is used:

"The order in which this function returns the file names is dependent on the file system type. With the NTFS file system and CDFS file systems, the names are usually returned in alphabetical order. With FAT file systems, the names are usually returned in the order the files were written to the disk, which may or may not be in alphabetical order. However, as stated previously, these behaviors are not guaranteed."

Considering this, we can't rely on the order. Using an additional data structure to store the path + size to sort and iterate it later may be a way to fix this.